allocator_test.go

```diff
@@ -1,23 +1,239 @@
+		require.True(t, isEachReplicaAssignedToDifferentNode(nodeAssignments, TestClusterSize))
+	}
+	require.True(t, nodeCapacityIsRespected(allocation, clusterCapacities, rangeDiskSpaceDemands))
+}
+
+func TestGIVENCapacityConstraintWithReplicationWHENInsufficientResourcesTHENAllocationFails(t *testing.T) {
+	const RangeSizeLimit = 5
+	const ReplicationFactor = 5
+	const TestClusterSize = 3
+	clusterCapacities := []int64{70, 80, 90}
+	rangeDiskSpaceDemands := []int64{25, 10, 12, 11, 10}
+	builtCluster := buildCluster(TestClusterSize, clusterCapacities)
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.False(t, status)
+	require.Nil(t, allocation)
+}
+
+func TestGIVENCapacityConstraintWHENInsufficientNodeCapacityTHENAllocationFails(t *testing.T) {
+	const RangeSizeLimit = 10
+	const ReplicationFactor = 1
+	const TestClusterSize = 3
+	builtCluster := buildCluster(TestClusterSize, []int64{70, 80, 80})
+	rangesToAllocate := make([]allocator.Range, RangeSizeLimit)
+	rangeDiskSpaceDemands := [RangeSizeLimit]int64{85, 75, 12, 11, 10, 9, 8, 7, 6, 6}
+	for initIndex := range rangesToAllocate {
+		rangesToAllocate[initIndex] = allocator.NewRange(
+			allocator.RangeID(initIndex),
+			ReplicationFactor,
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: rangeDiskSpaceDemands[initIndex]})
+	}
+	underTest := allocator.New(rangesToAllocate, builtCluster)
+	status, allocation := underTest.Allocate(allocator.WithNodeCapacityConstraint())
+	require.False(t, status)
+	require.Nil(t, allocation)
+}
+
+func buildCluster(clusterSize int64, nodeCapacities []int64) []allocator.Node {
+	cluster := make([]allocator.Node, clusterSize)
+	for index := 0; index < len(cluster); index++ {
+		cluster[index] = allocator.NewNode(
+			allocator.NodeID(index),
+			nil,
+			map[allocator.Resource]int64{allocator.DiskResource: nodeCapacities[index]})
+	}
+	return cluster
+}
+
+func nodeCapacitySupplier(clusterSize int64, minCapacity int64, maxCapacity int64) []int64 {
+	nodeCapacities := make([]int64, clusterSize)
+	for index := 0; index < len(nodeCapacities); index++ {
+		nodeCapacities[index] = generateRandomIntInRange(minCapacity, maxCapacity)
 	}
+	return nodeCapacities
+}
+
+func generateRandomIntInRange(lower int64, upper int64) int64 {
+	return rand.Int63n(upper-lower) + lower
+}
 
-	var ns []allocator.Node
-	for i := 0; i < 3; i++ {
-		ns = append(ns, allocator.NewNode(allocator.NodeID(i), nil, nil))
+func isValidNodeAssignment(nodeIds []allocator.NodeID, clusterSize int64) bool {
+	for index := 0; index < len(nodeIds); index++ {
```

Comment:
Why "index" when "i" suffices? Also, you can do `for _, id := range nodeIDs` to get the IDs directly (do the same below). Name the variable nodeIDs instead (do the same below).